### DISCUSSION TITLE
Home and End keys on MacOS

### DISCUSSION THREAD

**Discussion by binarylegit** [kovidgoyal/kitty#5248] [Q&A] [answered]
Hi, I'm really sorry if I'm missing something basic, I've tried my best to google/read the docs but I haven't been able to find a solution.

I'm on a Mac and am a newbie to using Kitty.  I'm used to using home and end keys to go to the beginning or end of the command I'm typing in terminals.  However, on MacOS running Kitty, the home and end keys seem to do nothing, other than a beep.

How can I get the home and end keys to send me to the beginning of the line or end of the line?

Thanks!

P.S. if this isn't the right place for this kind of question please let me know and feel free to delete it :).

**Comment by page-down** [comment]
> ... the home and end keys seem to do nothing, other than a beep.

You can troubleshoot the problem by checking the following.

- Update kitty to the latest version.
- Run `kitty --config=NONE --debug-input kitty +kitten show_key -m kitty` and check the debug log and the output of show_key kitten to confirm that the key is recognized correctly.
- See the FAQ to check TERM env and the TERMINFO database file.
- Troubleshoot the configuration of the shell, e.g. starting without any rc files.

**Reply by binarylegit** [reply]
I've uninstalled and re-installed to get the latest version.

Running the command you gave does appear that the correct keys are being recognized (if I'm reading this correctly):
```---------------- key down -------------------
NSEvent: type=KeyDown loc=(495.715,443.965) time=544.6 flags=0x800100 win=0x11deb0f00 winNum=269 ctxt=0x0 chars="" unmodchars="" repeat=0 keyCode=115
Press: native_key: 0x73 (<cc>) glfw_key: 0xe00c mods: none char_count: 1 deadKeyState: 0 repeat: 0 
	TextInputCtx: doCommandBySelector: (scrollToBeginningOfDocument:)
text: <none> glfw_key: HOME marked_text: ()
on_key_input: glfw key: 0xe00c native_code: 0x73 action: PRESS mods: none text: '' state: 0 updateIMEPosition: left=0.500000, top=13.000000, width=7.000000, height=12.500000
sent encoded key to child
----------------- key up --------------------
NSEvent: type=KeyUp loc=(495.715,443.965) time=544.7 flags=0x800100 win=0x11deb0f00 winNum=269 ctxt=0x0 chars="" unmodchars="" repeat=0 keyCode=115
Release: native_key: 0x73 (<cc>) glfw_key: 0xe00c mods: none 
on_key_input: glfw key: 0xe00c native_code: 0x73 action: RELEASE mods: none text: '' state: 0 updateIMEPosition: left=0.500000, top=50.500000, width=7.000000, height=12.500000
sent encoded key to child
---------------- key down -------------------
NSEvent: type=KeyDown loc=(495.715,443.965) time=545.6 flags=0x800100 win=0x11deb0f00 winNum=269 ctxt=0x0 chars="" unmodchars="" repeat=0 keyCode=119
Press: native_key: 0x77 (<cc>) glfw_key: 0xe00d mods: none char_count: 1 deadKeyState: 0 repeat: 0 
	TextInputCtx: doCommandBySelector: (scrollToEndOfDocument:)
text: <none> glfw_key: END marked_text: ()
on_key_input: glfw key: 0xe00d native_code: 0x77 action: PRESS mods: none text: '' state: 0 updateIMEPosition: left=0.500000, top=88.000000, width=7.000000, height=12.500000
sent encoded key to child
----------------- key up --------------------
NSEvent: type=KeyUp loc=(495.715,443.965) time=545.7 flags=0x800100 win=0x11deb0f00 winNum=269 ctxt=0x0 chars="" unmodchars="" repeat=0 keyCode=119
Release: native_key: 0x77 (<cc>) glfw_key: 0xe00d mods: none 
on_key_input: glfw key: 0xe00d native_code: 0x77 action: RELEASE mods: none text: '' state: 0 updateIMEPosition: left=0.500000, top=125.500000, width=7.000000, height=12.500000
sent encoded key to child
```

My TERM is xterm-kitty
and TERMINFO is /Applications/kitty.app/Contents/Resources/kitty/terminfo

I also added the .terminfo folder per the FAQ instructions to my home directory.

If I rename my .zshrc file so it isn't used the problem still persists.

**Comment by kovidgoyal** [comment, marked as answer]
This is what you need to educate zsh. Add it to your .zshrc.

```sh
# Setup bindings for both smkx and rmkx key variants
# Home
bindkey '\e[H'  beginning-of-line  
bindkey '\eOH'  beginning-of-line 
# End
bindkey '\e[F'  end-of-line
bindkey '\eOF'  end-of-line
# Up
bindkey '\e[A' up-line-or-beginning-search 
bindkey '\eOA' up-line-or-beginning-search 
# Down
bindkey '\e[B' down-line-or-beginning-search 
bindkey '\eOB' down-line-or-beginning-search
# Left
bindkey '\e[D' backward-char
bindkey '\eOD' backward-char
# Right
bindkey '\e[C' forward-char
bindkey '\eOC' forward-char
# Delete
bindkey '\e[3~' delete-char
# Backspace
bindkey '\e?' backward-delete-char
# PageUp
bindkey '\e[5~' up-line-or-history
# PageDown
bindkey '\e[6~' down-line-or-history
# Ctrl+Left
bindkey '\e[1;5D' backward-word
# Ctrl+Right
bindkey '\e[1;5C' forward-word
# Shift+Tab
bindkey '\e[Z' reverse-menu-complete
```

**Reply by binarylegit** [reply]
That did the trick, thank you so much Kovid!  I didn't realize it was a zsh issue

**Reply by page-down** [reply]
I noticed that in my environment zsh uses application mode by default.
The default zsh configuration contains all of the above key bindings, except for Home and End in normal mode.

```zsh
bindkey '\e[H' beginning-of-line
bindkey '\e[F' end-of-line
```

So having the line editor switch to application mode on entry, or binding these two keys solves the problem.

> If I rename my .zshrc file so it isn't used the problem still persists.

I'm not too familiar with zsh, not sure why it switched to normal mode, maybe it is affected by something else.

**Reply by mscottford** [reply]
I wanted to also support `Cmd+Left Arrow` and `Cmd+Right Arrow` so I added these lines:

```
# Cmd-Left
bindkey "\e[1;9D" beginning-of-line
# Cmd-Right
bindkey "\e[1;9C" end-of-line
```
